stackslib/src/net/api/getblock_v3.rs

```diff
@@ -0,0 +1,324 @@
+    pub index_block_hash: StacksBlockId,
+    /// consensus hash of this block (identifies its tenure; used by the tenure stream)
+    pub consensus_hash: ConsensusHash,
+    /// parent index block hash of the block to download (used by the tenure stream)
+    pub parent_block_id: StacksBlockId,
+    /// offset into the blob
+    pub offset: u64,
+    /// total number of bytes read.
+    pub total_bytes: u64,
+    /// Connection to the staging DB
+    pub staging_db_conn: NakamotoStagingBlocksConn,
+    /// rowid of the block
+    pub rowid: i64,
+}
+
+impl NakamotoBlockStream {
+    pub fn new(
+        chainstate: &StacksChainState,
+        block_id: StacksBlockId,
+        consensus_hash: ConsensusHash,
+        parent_block_id: StacksBlockId,
+    ) -> Result<Self, ChainError> {
+        let staging_db_path = chainstate.get_nakamoto_staging_blocks_path()?;
+        let db_conn = StacksChainState::open_nakamoto_staging_blocks(&staging_db_path, false)?;
+        let rowid = db_conn
+            .conn()
+            .get_nakamoto_block_rowid(&block_id)?
+            .ok_or(ChainError::NoSuchBlockError)?;
+
+        Ok(NakamotoBlockStream {
+            index_block_hash: block_id,
+            consensus_hash,
+            parent_block_id,
+            offset: 0,
+            total_bytes: 0,
+            staging_db_conn: db_conn,
+            rowid,
+        })
+    }
+
+    /// reset the stream to send another block.
+    /// Does not change the DB connection or consensus hash.
+    pub fn reset(
+        &mut self,
+        block_id: StacksBlockId,
+        parent_block_id: StacksBlockId,
+    ) -> Result<(), ChainError> {
+        let rowid = self
+            .staging_db_conn
+            .conn()
+            .get_nakamoto_block_rowid(&block_id)?
+            .ok_or(ChainError::NoSuchBlockError)?;
+
+        self.index_block_hash = block_id;
+        self.parent_block_id = parent_block_id;
+        self.offset = 0;
+        self.total_bytes = 0;
+        self.rowid = rowid;
+        Ok(())
+    }
+}
+
+/// Decode the HTTP request
+impl HttpRequest for RPCNakamotoBlockRequestHandler {
+    fn verb(&self) -> &'static str {
+        "GET"
+    }
+
+    fn path_regex(&self) -> Regex {
+        Regex::new(r#"^/v3/blocks/(?P<block_id>[0-9a-f]{64})$"#).unwrap()
+    }
+
+    /// Try to decode this request.
+    /// There's nothing to load here, so just make sure the request is well-formed.
+    fn try_parse_request(
+        &mut self,
+        preamble: &HttpRequestPreamble,
+        captures: &Captures,
+        query: Option<&str>,
+        _body: &[u8],
+    ) -> Result<HttpRequestContents, Error> {
+        if preamble.get_content_length() != 0 {
+            return Err(Error::DecodeError(
+                "Invalid Http request: expected 0-length body".to_string(),
+            ));
+        }
+
+        let block_id_str = captures
+            .name("block_id")
+            .ok_or(Error::DecodeError(
```

Comment:
Lazy evaluation will avoid unnecessary string allocations:

```suggestion
            .ok_or_else(|| Error::DecodeError(
```